jax/dtypes.py

```diff
@@ -105,14 +113,52 @@ def scalar_type_of(x):
   else:
     raise TypeError("Invalid scalar value {}".format(x))
 
+
+def _scalar_type_to_dtype(typ: type, value: Any = None):
+  """Return the numpy dtype for the given scalar type.
+
+  If `value` is specified and typ is int, then the resulting integer dtype will
+  be signed or unsigned depending on the value.
+
+  Examples
+  --------
+  >>> _scalar_dtype_to_dtype(int)
+  numpy.int64
+  >>> _scalar_dtype_to_dtype(float)
+  numpy.float64
+  >>> _scalar_dtype_to_dtype(complex)
+  numpy.complex128
+  >>> _scalar_dtype_to_dtype(int)
+  np.int64
+  >>> _scalar_dtype_to_dtype(int, 0)
+  np.int64
+  >>> _scalar_dtype_to_dtype(int, 1 << 63)
+  np.uint64
+  """
+  dtype = python_scalar_dtypes[typ]
+  # For integers, we return unsigned if required to store the value.
+  if typ is int and value is not None:
+    unsigned_dtype = _UINT_DTYPES[np.iinfo(dtype).bits]
+    if np.iinfo(dtype).min <= value <= np.iinfo(dtype).max:
+      pass
+    elif np.iinfo(unsigned_dtype).min < value <= np.iinfo(unsigned_dtype).max:
+      dtype = unsigned_dtype
+    else:
+      raise OverflowError(f"Python int {value} too large to convert to {dtype}")
+  return dtype
+
+
 def coerce_to_array(x):
   """Coerces a scalar or NumPy array to an np.array.
 
   Handles Python scalar type promotion according to JAX's rules, not NumPy's
   rules.
   """
-  dtype = python_scalar_dtypes.get(type(x), None)
-  return np.array(x, dtype) if dtype else np.array(x)
+  if type(x) in python_scalar_dtypes:
```

Comment:
What happens in cases like:

```
jnp.asarray([1 << 63])
```

`asarray` accepts, e.g., lists of Python scalars and I think this code will not handle them correctly.

Would it perhaps be possible to first let NumPy build an array and then cast it if needed?

Test this case?

Reply:
Good point... let me think on that a bit.

Reply:
One reason I didn't do that from the beginning is that `np.array(1 << 100)` returns dtype `object` rather than overflowing. From JAX's standpoint, in this case it's hard to know whether an array with object dtype implies an overflow error, or something different (e.g. `np.array([1, [1]])` also results in `dtype=object`)

Reply:
Hmm... this actually has enough weird corners that I might step back and write a brief design doc

Reply:
I fixed the list case here, and ensured that lists are no longer passed to `coerce_to_array`.